druid-shell/src/scale.rs

```diff
@@ -0,0 +1,388 @@
+    #[inline]
+    pub fn dp_to_px_y<T: Into<f64>>(&self, y: T) -> f64 {
+        y.into() * self.scale_y
+    }
+
+    /// Converts from display points into pixels,
+    /// using the x axis scale factor for `x` and the y axis scale factor for `y`.
+    #[inline]
+    pub fn dp_to_px_xy<T: Into<f64>>(&self, x: T, y: T) -> (f64, f64) {
+        (x.into() * self.scale_x, y.into() * self.scale_y)
+    }
+
+    /// Converts the `item` from display points into pixels,
+    /// using the x axis scale factor for coordinates on the x axis
+    /// and the y axis scale factor for coordinates on the y axis.
+    #[inline]
+    pub fn to_px<T: Scalable>(&self, item: &T) -> T {
+        item.to_px(self)
+    }
+
+    /// Converts from pixels into display points, using the x axis scale factor.
+    #[inline]
+    pub fn px_to_dp_x<T: Into<f64>>(&self, x: T) -> f64 {
+        x.into() / self.scale_x
+    }
+
+    /// Converts from pixels into display points, using the y axis scale factor.
+    #[inline]
+    pub fn px_to_dp_y<T: Into<f64>>(&self, y: T) -> f64 {
+        y.into() / self.scale_y
+    }
+
+    /// Converts from pixels into display points,
+    /// using the x axis scale factor for `x` and the y axis scale factor for `y`.
+    #[inline]
+    pub fn px_to_dp_xy<T: Into<f64>>(&self, x: T, y: T) -> (f64, f64) {
+        (x.into() / self.scale_x, y.into() / self.scale_y)
+    }
+
+    /// Converts the `item` from pixels into display points,
+    /// using the x axis scale factor for coordinates on the x axis
+    /// and the y axis scale factor for coordinates on the y axis.
+    #[inline]
+    pub fn to_dp<T: Scalable>(&self, item: &T) -> T {
+        item.to_dp(self)
+    }
+}
+
+impl Scalable for Vec2 {
+    /// Converts a `Vec2` from display points into pixels,
+    /// using the x axis scale factor for `x` and the y axis scale factor for `y`.
+    #[inline]
+    fn to_px(&self, scale: &Scale) -> Vec2 {
+        Vec2::new(self.x * scale.scale_x, self.y * scale.scale_y)
+    }
+
+    /// Converts a `Vec2` from pixels into display points,
+    /// using the x axis scale factor for `x` and the y axis scale factor for `y`.
+    #[inline]
+    fn to_dp(&self, scale: &Scale) -> Vec2 {
+        Vec2::new(self.x / scale.scale_x, self.y / scale.scale_y)
+    }
+}
+
+impl Scalable for Point {
+    /// Converts a `Point` from display points into pixels,
+    /// using the x axis scale factor for `x` and the y axis scale factor for `y`.
+    #[inline]
+    fn to_px(&self, scale: &Scale) -> Point {
+        Point::new(self.x * scale.scale_x, self.y * scale.scale_y)
+    }
+
+    /// Converts a `Point` from pixels into display points,
+    /// using the x axis scale factor for `x` and the y axis scale factor for `y`.
+    #[inline]
+    fn to_dp(&self, scale: &Scale) -> Point {
+        Point::new(self.x / scale.scale_x, self.y / scale.scale_y)
+    }
+}
+
+impl Scalable for Line {
```

Comment:
Is this used? I would have thought that scaling is happening to points, sizes, and rects, but not other shapes.